components/raftstore-v2/src/operation/query/lease.rs

```diff
@@ -291,4 +295,64 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         }
         state
     }
+
+    // If lease expired, we will send a noop read index to renew lease.
+    fn try_renew_leader_lease<T>(&mut self, ctx: &mut StoreContext<EK, ER, T>) {
+        debug!(self.logger,
+            "renew lease";
+            "region_id" => self.region_id(),
+            "peer_id" => self.peer_id(),
+        );
+
+        let current_time = *ctx.current_time.get_or_insert_with(monotonic_raw_now);
+        if self.need_renew_lease_at(ctx, current_time) {
+            let mut cmd = new_read_index_request(
+                self.region_id(),
+                self.region().get_region_epoch().clone(),
+                self.peer().clone(),
+            );
+            cmd.mut_header().set_read_quorum(true);
+            let (ch, _) = QueryResChannel::pair();
+            self.read_index(ctx, cmd, ch);
+        }
+    }
+
+    fn need_renew_lease_at<T>(
+        &self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        current_time: Timespec,
+    ) -> bool {
+        let renew_bound = match self.leader_lease().need_renew(current_time) {
+            Some(ts) => ts,
+            None => return false,
+        };
+        let max_lease = ctx.cfg.raft_store_max_leader_lease();
+        let has_overlapped_reads = self.pending_reads().back().map_or(false, |read| {
+            // If there is any read index whose lease can cover till next heartbeat
+            // then we don't need to propose a new one
+            read.propose_time + max_lease > renew_bound
+        });
+        let has_overlapped_writes = self.proposals().back().map_or(false, |proposal| {
+            // If there is any write whose lease can cover till next heartbeat
+            // then we don't need to propose a new one
+            proposal
+                .propose_time
+                .map_or(false, |propose_time| propose_time + max_lease > renew_bound)
+        });
+        !has_overlapped_reads && !has_overlapped_writes
+    }
+}
+
+impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER, T> {
+    fn register_check_leader_lease_tick(&mut self) {
+        self.schedule_tick(PeerTick::CheckLeaderLease)
+    }
+
+    pub fn on_check_leader_lease_tick(&mut self) {
+        if !self.fsm.peer_mut().is_leader() {
+            return;
+        }
+        self.fsm.peer_mut().try_renew_leader_lease(self.store_ctx);
+        self.register_check_leader_lease_tick();
```

Comment:
should this register_check_leader_lease_tick be called when the peer is elected as leader?